widget/color.go

```diff
@@ -0,0 +1,254 @@
+	}
+	return box
+}
+
+func newCheckeredBackground() *canvas.Raster {
+	return canvas.NewRasterWithPixels(func(x, y, _, _ int) color.Color {
+		const boxSize = 25
+
+		if (x/boxSize)%2 == (y/boxSize)%2 {
+			return color.Gray{Y: 58}
+		}
+
+		return color.Gray{Y: 84}
+	})
+}
+
+const (
+	preferenceRecents    = "color_recents"
+	preferenceMaxRecents = 8
+)
+
+func readRecentColors() (recents []string) {
+	for _, r := range strings.Split(fyne.CurrentApp().Preferences().String(preferenceRecents), ",") {
+		if r != "" {
+			recents = append(recents, r)
+		}
+	}
+	return
+}
+
+func writeRecentColor(color string) {
+	recents := []string{color}
+	for _, r := range readRecentColors() {
+		if r == color {
+			continue // Color already in recents
+		}
+		recents = append(recents, r)
+	}
+	if len(recents) > preferenceMaxRecents {
+		recents = recents[:preferenceMaxRecents]
+	}
+	fyne.CurrentApp().Preferences().SetString(preferenceRecents, strings.Join(recents, ","))
+}
+
+func colorToString(c color.Color) string {
+	var red, green, blue, alpha uint8
+	switch col := c.(type) {
+	case *color.NRGBA:
+		red, green, blue, alpha = col.R, col.G, col.B, col.A
+	default:
+		r, g, b, a := c.RGBA() // TODO FIXME this returns alpha-pre-multiplied
+		// TODO FIXME possible loss of precision, should this be x/65535*255?
+		red = uint8(r)
+		green = uint8(g)
+		blue = uint8(b)
+		alpha = uint8(a)
+	}
+	if alpha == 0xff {
+		return fmt.Sprintf("#%02x%02x%02x", red, green, blue)
+	}
+	return fmt.Sprintf("#%02x%02x%02x%02x", red, green, blue, alpha)
+}
+
+func stringToColor(s string) color.Color {
+	var c color.NRGBA
+	var err error
+	if len(s) == 7 {
+		c.A = 0xFF
+		_, err = fmt.Sscanf(s, "#%02x%02x%02x", &c.R, &c.G, &c.B)
+	} else {
+		_, err = fmt.Sscanf(s, "#%02x%02x%02x%02x", &c.R, &c.G, &c.B, &c.A)
+	}
+	if err != nil {
+		fyne.LogError("Couldn't parse color:", err)
+	}
+	return c
+}
+
+func stringsToColors(ss ...string) (colors []color.Color) {
+	for _, s := range ss {
+		if s == "" {
+			continue
+		}
+		colors = append(colors, stringToColor(s))
+	}
+	return
+}
+
+func colorToHSLA(c color.Color) (float64, float64, float64, float64) {
+	r, g, b, a := colorToRGBA(c)
+	h, s, l := rgbToHsl(r, g, b)
+	return h, s, l, a
+}
+
+func colorToRGBA(c color.Color) (r, g, b, a float64) {
+	switch col := c.(type) {
+	case *color.NRGBA:
+		// Convert to range 0.0-1.0
+		r = float64(col.R) / 255.0
+		g = float64(col.G) / 255.0
+		b = float64(col.B) / 255.0
+		a = float64(col.A) / 255.0
+	default:
+		red, green, blue, alpha := c.RGBA() // TODO FIXME this returns alpha-pre-multiplied
+		// Convert to range 0.0-1.0
+		r = float64(red) / 65535.0
+		g = float64(green) / 65535.0
+		b = float64(blue) / 65535.0
+		a = float64(alpha) / 65535.0
+	}
+	return
+}
+
+func cartesianToPolar(x, y, width, height float64) (angle, radius, limit float64) {
+	dx := float64(x - (width / 2))
+	dy := float64(y - (height / 2))
+	angle = math.Atan2(dy, dx)
+	radius = math.Sqrt(dx*dx + dy*dy)
```

Comment:
Should we perhaps be using `complex128` instead and just use the `cmplx.Polar()` function from the standard library? https://pkg.go.dev/math/cmplx?tab=doc#Polar

Reply:
It is a little weird to say X is real, and Y is imaginary but in general it is better to use a well-tested standard library